### ISSUE TITLE
Assert using LocalManualEvent

### ISSUE THREAD

**Issue by tchaloupka** [vibe-d/vibe-core#37]
I'm using the `createManualEvent` to signal waiting tasks for new data.
But with vibe-core I sometimes get:

```
core.exception.AssertError@../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/task.d(784): Calling task is running, but scheduled to be resumed!?
----------------
??:? _d_assert_msg [0x13f404e]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/task.d:784 nothrow @safe void vibe.core.task.TaskScheduler.switchTo(vibe.core.task.Task, std.typecons.Flag!("defer").Flag) [0x1388a25]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/core.d:700 nothrow @safe void vibe.core.core.switchToTask(vibe.core.task.Task) [0x13a6195]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/internal/async.d-mixin-113:121 nothrow @safe void vibe.internal.async.asyncAwaitAny!(true, "vibe.core.sync.ThreadLocalWaiter!false.ThreadLocalWaiter.wait!true.wait", vibe.internal.async.Waitable!(void delegate() nothrow @safe, vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.wait!(true).wait(core.time.Duration, eventcore.driver.__T6HandleVAyaa5_6576656e74TS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleVS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleS1N1Z.Handle, bool delegate() nothrow @safe).__lambda5, vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.wait!(true).wait(core.time.Duration, eventcore.driver.__T6HandleVAyaa5_6576656e74TS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleVS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleS1N1Z.Handle, bool delegate() nothrow @safe).__lambda6).Waitable).asyncAwaitAny(ref vibe.internal.async.Waitable!(void delegate() nothrow @safe, vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.wait!(true).wait(core.time.Duration, eventcore.driver.__T6HandleVAyaa5_6576656e74TS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleVS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleS1N1Z.Handle, bool delegate() nothrow @safe).__lambda5, vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.wait!(true).wait(core.time.Duration, eventcore.driver.__T6HandleVAyaa5_6576656e74TS9eventcore6driver28__T6HandleVAyaa2_6664TmVmN1Z6HandleVS9eventcore6driver28__T6HandleV../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/internal/async.d-mixin-113:121 /rt/backtrace
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/sync.d:1376 nothrow @safe void vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.emitWaiter(vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.TaskWaiter*) [0x137fa10]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/sync.d:1359 nothrow @safe int vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.emit().__foreachbody2(vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.TaskWaiter*) [0x137f8e3]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/internal/list.d:53 nothrow @safe int vibe.internal.list.CircularDList!(vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.TaskWaiter*).CircularDList.opApply(int delegate(vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.TaskWaiter*) nothrow @safe) [0x139a755]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/sync.d:1357 nothrow @safe void vibe.core.sync.ThreadLocalWaiter!(false).ThreadLocalWaiter.emit() [0x137f86e]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/sync.d:721 nothrow @safe int vibe.core.sync.LocalManualEvent.emit() [0x1377f7e]
../../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.7/vibe-mqtt/source/mqttd/client.d:369 @safe ushort mqttd.client.SessionQueue!(1).SessionQueue.add(mqttd.client.MessageContext) [0x134df77]
../../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.7/vibe-mqtt/source/mqttd/client.d:315 @safe ushort mqttd.client.SessionQueue!(1).SessionQueue.add(mqttd.messages.Publish, mqttd.client.PacketState, mqttd.client.PacketOrigin) [0x134dcb6]
../../../../../.dub/packages/vibe-mqtt-0.2.0-alpha.7/vibe-mqtt/source/mqttd/client.d:674 @safe void mqttd.client.MqttClient.publish!(immutable(ubyte)[]).publish(const(immutable(char)[]), const(immutable(ubyte)[]), mqttd.messages.QoSLevel, bool) [0x1153bbc]
../common/mq.d:97 @trusted void common.mq.mqttSender() [0x11537f6]
../common/mq.d:60 @safe void common.mq.initMq(immutable(char)[], void delegate(immutable(char)[], immutable(ubyte)[]) @safe, const(immutable(char)[][])...).__lambda6() [0x1153740]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/task.d:537 void vibe.core.task.TaskFuncInfo.set!(void function() @safe*).set(ref void function() @safe*).callDelegate(ref vibe.core.task.TaskFuncInfo) [0x1060b02]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/task.d:567 void vibe.core.task.TaskFuncInfo.call() [0x138814d]
../../../../../.dub/packages/vibe-core-1.2.0/vibe-core/source/vibe/core/task.d:380 nothrow void vibe.core.task.TaskFiber.run() [0x1387579]
??:? void core.thread.Fiber.run() [0x14511bb]
??:? fiber_entryPoint [0x1451096]
??:? [0xffffffff]
```

I don't understand the message if this can be a problem on our side or with vibe-core?

**Comment by MartinNowak** [comment]
As assertion is usually an internal bug, a bit more context (code) would be helpful to reproduce the problem though.

**Comment by s-ludwig** [comment]
It would also be interesting to know if there may have been any uncaught exceptions in a task. I've observed that this currently may leave the task scheduling queue in an invalid state.

**Comment by tchaloupka** [comment]
I tried to run it with GDB, but no success to detect anything unusual. Output is also free of some uncaught errors.
Unfortunately it's seems to be not too easy to reproduce it in the simple example. I tried to write from scratch some task messaging with `LocalManualEvent` to simulate the behavior but without success.
I've also tried to run just `vibe-mqtt` publisher example code to burst messages, but also without successful reproduction of the problem.

I can trigger it on the service in about 20 of same REST API requests mixed with cca 2 other ones. Strangely not with the `curl` yet.
It goes like: Service A calls REST API on Service B, which then sends MQTT message to message broker.
